ruby/lib/extism.rb

```diff
@@ -74,5 +91,9 @@ def call(name, data, &block)
       buf = C.extism_output_get(@plugin)
       return block.call(buf, out_len)
     end
+    
+    def destroy()
+      C.extism_plugin_destroy(@plugin)
+    end
```

Comment:
This comment applies to any languages with GC, but since we initialize a plugin when we initialize the object, should we also destroy the plugin when the object gets freed by the GC? In ruby these are called "finalizers". In python i think there is also a way to hook into this. We could also use context managers.

Reply:
The more I think about it, the more I think it's maybe a bad idea. Because there could be cases where the GC collects an object but the plugin isn't done working and it would be a very tricky thing to debug. It's probably better to keep this explicit and just make sure that people understand that they are allocating this plugin and may need to deallocate it themselves explicitly if they are creating many of them.

Reply:
I wasn’t able to find information about finalizers in Ruby. But I think it would be ideal to use finalizers when possible and if we find out there is some kind of lifetime issue we can fix that - I don’t think it will be an issue in practice.

In Python we’re using the __del__ method to do the cleanup.

Reply:
> What is the typical way of doing finalization in Ruby?

The semantics for this in ruby a little weird. I see in python that the GC eventually will call `__del__` for you if the reference count goes to 0.
In ruby you can define a finalizer in the initializer:

```ruby
class Plugin  
  def initialize(wasm, wasi=false, config=nil)
    # ...
    ObjectSpace.define_finalizer(self, proc { |_id| self.destroy } )
  end
end
```

> But I think it would be ideal to use finalizers when possible and if we find out there is some kind of lifetime issue we can fix that - I don’t think it will be an issue in practice.

My concern would be someone say creating a plugin and firing off a call to it from a web-request. then the plugin object goes out of scope and gets garbage collected when the request is done (but maybe the plugin is not done). It does seem like there is no asynchronous way to do a call though. So maybe that's not a problem since call has to return (meaning work is done). I've yet to observe or test this just thinking out loud.

Reply:
Ah that is an interesting point, I wasn’t thinking about the possibility of async plugins. On the Rust side the plug-in is protected by a mutex so I don’t think that can actually happen. Either way that can be something we handle in the Rust code.

I will try adding that to the Ruby SDK, thanks for the example!

Reply:
I kept getting warnings about the finalizer referencing the object being finalized, but I found a slightly more complex way that works using a global hash from object_ids to plugin ids